### PR TITLE
[FAI-379] Implement flag for when no counterfactual was found

### PR DESCRIPTION
**Pull request by ruivieira**
The counterfactual search will always return a result (the "best" solution) even when no valid counterfactual was found.
This implements a behaviour where if the "hard" constraints are not met, the counterfactual result can still be checked for validity. The changes consists of:

- Including the OptaPlanner score in the `CounterfactualResult`
- Adding a method `CounterfactualResult::isValid` which indicated whether the counterfactual is valid or not (by checking under the hood the OptaPlanner `Score::isFeasible`)

> Many thanks for submitting your Pull Request :heart:! 
> 
> Please make sure that your PR meets the following requirements:
> 
> - [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
> - [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
> - [ ] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
> - [x] Pull Request contains link to the JIRA issue
> - [ ] Pull Request contains link to any dependent or related Pull Request
> - [x] Pull Request contains description of the issue
> - [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualResult.java [hunk 1]
```diff
@@ -42,4 +46,8 @@ public CounterfactualResult(List<CounterfactualEntity> entities, List<Prediction
     public List<PredictionOutput> getOutput() {
         return output;
     }
+
+    public boolean isValid() {
+        return score.isFeasible();
```
Comment (manstis) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualResult.java]: If you only need score for `isFeasible` can we not pass a boolean to the constructor, remove the import of `BendableBigDecimalScore` and hence the coupling of Counter Factuals from OptaPlanner? (Or at least the _public_ result from the internal implementation of the algorithm that uses OptaPlanner).

Comment (ruivieira) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualResult.java]: Good point!

**Review by danielezonca — APPROVED** [review]
LGTM 👍

**Review by r00ta — APPROVED** [review]

**Comment by r00ta** [comment]
Jenkins retest this please

Hi there, we just added a maven plugin to check the codestyle. I retrigger CI because of that

**Comment by ruivieira** [comment]
Jenkins retest this please

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java [hunk 2]
```diff
@@ -501,5 +505,48 @@ void testCounterfactualMatchNoThreshold(int seed) throws ExecutionException, Int
...
         assertTrue(predictionScore < 0.1);
+        assertTrue(result.isValid());
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = { 0, 1, 2, 3, 4 })
+    void testNoCounterfactualPossible(int seed)
+            throws ExecutionException, InterruptedException, TimeoutException {
+        Random random = new Random();
+        random.setSeed(seed);
```
Comment (r00ta) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: where is this random used?

Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @r00ta well spotted! Not directly in the test body, but via OP. In any case, as @danielezonca suggested, added a `PerturbationContext` and feature perturbations using the fixed random seeds.

**Comment by ruivieira** [comment]
@r00ta @danielezonca comments addressed, ptal.

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java (hunk 2 above)
Comment (r00ta) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: Hi @ruivieira Is there any update on the topic we discussed in chat? Meaning setting the OP env variable to have reproducible tests?

Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @r00ta we are using the `REPRODUCIBLE` mode, meaning we just need to implement a random seed setter in the CF builder for OP using [SolverConfig](https://docs.optaplanner.org/latest/optaplanner-javadoc/org/optaplanner/core/config/solver/SolverConfig.html#setRandomSeed(java.lang.Long)). Would that be in scope for this PR, or open a new one?

Comment (r00ta) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @ruivieira sorry, I didn't get it. If we are already using it, why do we have to create a ticket?

Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @r00ta After checking with @triceo (cc) if we want to make sure OP uses the random seed specified in the unit test, it is best to pass it explicitly to OP using the above method. Should I change the unit tests to pass the seed directly to the `SolverConfig`?

Comment (r00ta) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @ruivieira , yes, that would be great thanks. Feel free to create another ticket for that

Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @r00ta The random seed is now explicitly passed to OP for reproducibility, in https://github.com/kiegroup/kogito-apps/pull/674/commits/e4e07ef901953baa372f1e936aba15ee65ea17aa

Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:539]: @r00ta didn't see your last message. done, in any case :)

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java [hunk 3]
```diff
@@ -49,14 +48,15 @@
...
-    private CounterfactualResult runCounterfactualSearch(List<Output> goal,
+    private CounterfactualResult runCounterfactualSearch(Long randomSeed, List<Output> goal,
             List<Boolean> constraints,
             DataDomain dataDomain,
             List<Feature> features,
             PredictionProvider model) throws InterruptedException, ExecutionException, TimeoutException {
         final TerminationConfig terminationConfig = new TerminationConfig().withScoreCalculationCountLimit(steps);
         final SolverConfig solverConfig = CounterfactualConfigurationFactory
                 .builder().withTerminationConfig(terminationConfig).build();
+        solverConfig.setRandomSeed(randomSeed);
```
Comment (triceo) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:79]: If you want to make absolutely sure this is used and that the work done is completely reproducible, also set `EnvironmentMode` to `REPRODUCIBLE`.

Comment (ruivieira) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainerTest.java:79]: @triceo Thanks! Added with https://github.com/kiegroup/kogito-apps/pull/674/commits/9e97780e0a0cfd21c7301adfb0ccd1bfa2d3023e

**Review by jiripetrlik — APPROVED** [review]

**Pull request merged**